zebra-chain/src/orchard/arbitrary/keys.rs

```diff
@@ -0,0 +1,255 @@
+    }
+}
+
+impl fmt::Debug for IncomingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("IncomingViewingKey")
+            .field("dk", &self.dk)
+            .field("ivk", &self.ivk)
+            .finish()
+    }
+}
+
+impl fmt::Display for IncomingViewingKey {
+    /// The _raw encoding_ of an **Orchard** _incoming viewing key_.
+    ///
+    /// <https://zips.z.cash/protocol/protocol.pdf#orchardfullviewingkeyencoding>
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&hex::encode(<[u8; 64]>::from(*self)))
+    }
+}
+
+impl Eq for IncomingViewingKey {}
+
+impl From<IncomingViewingKey> for [u8; 64] {
+    fn from(ivk: IncomingViewingKey) -> [u8; 64] {
+        let mut bytes = [0u8; 64];
+
+        bytes[..32].copy_from_slice(&<[u8; 32]>::from(ivk.dk));
+        bytes[32..].copy_from_slice(&<[u8; 32]>::from(ivk.ivk));
+
+        bytes
+    }
+}
+
+impl TryFrom<[u8; 64]> for IncomingViewingKey {
+    type Error = &'static str;
+
+    /// Convert an array of bytes into a [`IncomingViewingKey`].
+    ///
+    /// Returns an error if the encoding is malformed or if it [encodes the scalar additive
+    /// identity, 0][1].
+    ///
+    /// > ivk MUST be in the range {1 .. 𝑞P - 1}
+    ///
+    /// [1]: https://zips.z.cash/protocol/protocol.pdf#orchardinviewingkeyencoding
+    fn try_from(bytes: [u8; 64]) -> Result<Self, Self::Error> {
+        let mut dk_bytes = [0u8; 32];
+        dk_bytes.copy_from_slice(&bytes[..32]);
+        let dk = DiversifierKey::from(dk_bytes);
+
+        let mut ivk_bytes = [0u8; 32];
+        ivk_bytes.copy_from_slice(&bytes[32..]);
+
+        let possible_scalar = pallas::Scalar::from_repr(ivk_bytes);
+
+        if possible_scalar.is_some().into() {
+            let scalar = possible_scalar.unwrap();
+            if scalar.is_zero().into() {
+                Err("pallas::Scalar value for Orchard IncomingViewingKey is 0")
+            } else {
+                Ok(Self {
+                    dk,
+                    ivk: possible_scalar.unwrap(),
+                })
+            }
+        } else {
+            Err("Invalid pallas::Scalar value for Orchard IncomingViewingKey")
+        }
+    }
+}
+
+impl TryFrom<FullViewingKey> for IncomingViewingKey {
+    type Error = &'static str;
+
+    /// Commit^ivk_rivk(ak, nk) :=
+    ///     SinsemillaShortCommit_rcm(︁
+    ///        "z.cash:Orchard-CommitIvk",
+    ///        I2LEBSP_l^Orchard_base(ak) || I2LEBSP_l^Orchard_base(nk)︁
+    ///     ) mod r_P
+    ///
+    /// <https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents>
+    /// <https://zips.z.cash/protocol/nu5.pdf#concreteprfs>
+    #[allow(non_snake_case)]
+    #[allow(clippy::unwrap_in_result)]
+    fn try_from(fvk: FullViewingKey) -> Result<Self, Self::Error> {
+        let mut M: BitVec<u8, Lsb0> = BitVec::new();
+
+        // I2LEBSP_l^Orchard_base(ak)︁
+        //
+        // Converting a Point to bytes and back is infalliable.
```

Comment:
I added this comment, is it correct?

Reply:
Yep, this is correct, there's just a typo. I think this piece of code can be ultimately removed, though

```suggestion
        // Converting a Point to bytes and back is infallible.
```

Reply:
Do you think it's worth removing this code now? I don't mind either way.

Reply:
I think so, I don't see a reason to keep it if we want to make things easier for auditors.

Reply:
It seems like I can remove this easily, I'll push a commit.